Tutorial edits (#585)
MicrosoftDocs/quantum-docs-pr: articles/quickstart.md

---
# Write your first quantum program

In this QuickStart, we show you how to write your first Q# program that manipulates and measures qubits and demonstrates the effects of superposition and entanglement.  This guides you on installing the QDK, building the program and executing that program on a quantum simulator.  

You will write an application called Bell to demonstrate quantum entanglement.  The name Bell is in reference to Bell states, which are specific quantum states of two qubits that are used to represent the simplest examples of superposition and quantum entanglement. 

## Pre-requisites

If you are ready to start coding, follow these steps before proceeding: 

* [Install](xref:microsoft.quantum.install) the Quantum Development Kit using your preferred language and development environment
* If you already have the QDK installed, make sure you have [updated](xref:microsoft.quantum.update) to the latest version

You can also follow along with the narrative without installing the QDK, reviewing the overviews of the Q# programming language and the first concepts of quantum computing.

## Demonstrating qubit behavior with Q#

> Recall our simple [definition of a qubit](xref:microsoft.quantum.overview.what#the-qubit).  Where classical bits hold a single binary value such as a 0 or 1, the state of a qubit can be in a **superposition** of 0 and 1 simultaneously.  Conceptually, a qubit can be thought of as a direction in space (also known as a vector).  A qubit can be in any of the possible directions. The two **classical states** are the two directions; representing 100% chance of measuring 0 and 100% chance of measuring 1.  This representation is also more formally visualized by the [block sphere]

Reviewer: ```suggestion
Recall our simple [definition of a qubit](xref:microsoft.quantum.overview.what#the-qubit).  Where classical bits hold a single binary value such as a 0 or 1, the state of a qubit can be in a **superposition** of 0 and 1.  Conceptually, a qubit can be thought of as a direction in space (also known as a vector).  A qubit can be in any of the possible directions. The two **classical states** are the two directions; representing 100% chance of measuring 0 and 100% chance of measuring 1.  This representation is also more formally visualized by the [block sphere]
```

Reviewer: ```suggestion
Recall our simple [definition of a qubit](xref:microsoft.quantum.overview.what#the-qubit).  Where classical bits hold a single binary value such as a 0 or 1, the state of a qubit can be in a **superposition** of 0 and 1 simultaneously.  Conceptually, a qubit can be thought of as a direction in space (also known as a vector).  A qubit can be in any of the possible directions. The two **classical states** are the two directions; representing 100% chance of measuring 0 and 100% chance of measuring 1.  This representation is also more formally visualized by the [Bloch sphere]
```

Author: For the first, I would like to keep "simultaneously", as noted above.  I don't see what the change you suggest in the second, but I did correct the "block sphere" link

## Write a Q# operation

> Our goal is to prepare two qubits in a [Bell state](https://en.wikipedia.org/wiki/Bell_state) showing entanglement.  Our goal is to prepare two qubits in a specific quantum state, demonstrating how to operate on qubits with Q# to change their state and demonstrate the effects of superposition and entanglement. We will build this up piece by piece to demonstrate qubit states, operations, and measurement.

Reviewer: This sounds redundant at the moment?

Author: correcting.

> **Overview:**  In the first code below, we show you how to work with qubits in Q#.  We’ll introduce two operations, M and X that transform the state of a qubit. 

Reviewer: ```suggestion
**Overview:**  In the first code below, we show you how to work with qubits in Q#.  We’ll introduce two operations, [M](xref:microsoft.quantum.intrinsic.m) and [X](xref:microsoft.quantum.intrinsic.x) that transform the state of a qubit. 
```

Author: Normally, I'm a bit fan of hyperlinks to further detail.  But not in this first leaner article.  The link does not lead to a page that will further inform the new learner.  It's not time to introduce them to the formal definitions of operations, imho.

> In this code snippet, an operation `Set` is defined that takes as a parameter a qubit and another parameter, `desired`, representing the state that we would like the qubit to be in.  The operation `Set` performs a measurement on the qubit using the operation `M`.  In Q#, a qubit measurement always returns either  `Zero` or `One`.  If the measurement returns a value not equal to a desired value, Set “flips” the qubit; that is, it executes an `X` operation, which changes the qubit state to a new state in which the probabilities of a measurement returning `Zero` and `One` are reversed.  To demonstrate the effect of the `Set` operation, a `TestBellState` operation is then added.  This operation takes as input a `Zero` or `One`, and calls the `Set` operation some number of times with that input, and counts the number of times that `Zero` was returned from the measurement of the qubit and the number of times that `One` was returned. Of course, in this first simulation of the `TestBellState` operation, we expect that the output will show that all measurements of the qubit set with `Zero` as the parameter input will return `Zero`, and all measurements of a qubit set with `One` as the parameter input will return `One`.  Further on, we’ll add code to `TestBellState` to demonstrating superposition and entanglement.

Reviewer: In some places, we refer to the `X` operation as `` `X` `` (that is, typeset as an inline code block), while in others we use `X` (that is, typeset in roman as plain prose).

Author: I found one instance of X and M without the quote marks.  I'll upate.

    }
    ```

    This operation may now be called to set a qubit to a classical state, either returning `Zero` 100% of the time or returning `One` 100% of the time.  `Zero` and `One` are constants that represent the only two possible results of a measurement of a qubit.

>     The operation `Set` measures the qubit, if it's in the state we want, we leave it alone, otherwise, by executing the `X` operation, we change the qubit state to the desired state.

Reviewer: This sentence feels a bit awkward at the moment. I apologize that I can't quite put my finger on why, though.

Author: Don't want to say "flip", because new learner does not have conceptual image of qubit where "flipping" is understood.

Reviewer 2: It works for me, I agree it's more intuitive than "flip"

Reviewer: To be sure, wasn't suggestion going back to flip. Maybe it would help to separate into two sentences?
```suggestion
    The operation `Set` measures the qubit.
    If the qubit is in the state we want, `Set` leaves it alone; otherwise, by executing the `X` operation, we change the qubit state to the desired state.
```

### About Q# operations

> A Q# operation is a quantum subroutine. That is, it is a callable routine that contains quantum operations.

Reviewer: I might suggest another word than "contains" here, perhaps it might be more clear to say that a Q# operation can allocate qubits and call other Q# operations, including the intrinsic operations such as `X` and `M` introduced above?

Author: For a new learner, "contains" is better, I think, than the clearer definition you propose.  Also, this is verbatim from the definition of Q# operation in the language doc.

The arguments to an operation are specified as a tuple, within parentheses.

The return type of the operation is specified after a colon. In this case, the `Set` operation has no return, so it is marked as returning `Unit`. This is the Q# equivalent of `unit` in F#, which is roughly analogous to `void` in C#, and an empty tuple (`Tuple[()]`) in Python.

You have used two quantum operations in your first Q# operation:

* The [M](xref:microsoft.quantum.intrinsic.m) operation, which measures the state of the qubit
* The [X](xref:microsoft.quantum.intrinsic.x) operation, which flips the state of a qubit

> A quantum operation transforms the state of a qubit. Sometime people talk about quantum gates instead of operations, in analogy to classical logic gates. This is rooted in the early days of quantum computing when algorithms were merely a theoretical construct and visualized as diagrams similarly to circuit diagrams in classical computing.

Reviewer: 👍

Author: Thumbs up to Bettina for the suggested prose!

## Prepare superposition

> **Overview** Now let’s look at how Q# expresses ways to put qubits in superposition.  Recall that the state of a qubit can be in a superposition of 0 and 1.  We’ll use the `Hadamard` operation to accomplish this. If the qubit is in either of the classical states (where a measurement returns zero always or one always), then the `Hadamard` or `H` operation will put the qubit in a state where a measurement of the qubit will return `Zero` 50% of the time and return `One` 50% of the time.  Conceputually, the qubit can be thought of as halfway between the `Zero` and `One`.  Now, when we simulate the `TestBellState` operation, we will see the results will return roughly an equal number of `Zero` and `One` after measurement.  

Reviewer 2: ```suggestion
**Overview** Now let’s look at how Q# expresses ways to put qubits in superposition.  Recall that the state of a qubit can be in a superposition of 0 and 1.  We’ll use the `Hadamard` operation to accomplish this. If the qubit is in either of the classical states (where a measurement returns `Zero` always or `One` always), then the `Hadamard` or `H` operation will put the qubit in a state where a measurement of the qubit will return `Zero` 50% of the time and return `One` 50% of the time.  Conceputually, the qubit can be thought of as halfway between the `Zero` and `One`.  Now, when we simulate the `TestBellState` operation, we will see the results will return roughly an equal number of `Zero` and `One` after measurement.  
```

Reviewer 2: I don't think it's necessary before Ignite, but this is a place where it might be useful to include some of the Bloch sphere visuals Eduardo added to QRNG

> First we'll just try to flip the qubit. This is accomplished by performing an `X` operation before we measure it in `TestBellState`:

Reviewer 2: ```suggestion
First we'll just try to flip the qubit (if the qubit is in `Zero` state will flip to `One` and vice versa). This is accomplished by performing an `X` operation before we measure it in `TestBellState`:
```
